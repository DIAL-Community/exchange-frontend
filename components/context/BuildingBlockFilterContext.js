import React, { createContext, useState } from 'react'

const BuildingBlockFilterContext = createContext()
const BuildingBlockFilterDispatchContext = createContext()

const BuildingBlockFilterProvider = ({ children }) => {
  const [showMature, setShowMature] = useState(false)
  const [sdgs, setSDGs] = useState([])
  const [useCases, setUseCases] = useState([])
  const [workflows, setWorkflows] = useState([])
  const [categoryTypes, setCategoryTypes] = useState([])

  const [search, setSearch] = useState('')
  const [sortColumn, setSortColumn] = useState('name')
  const [sortDirection, setSortDirection] = useState('asc')

  const setSdgs = setSDGs

  const buildingBlockFilterValues = {
    showMature, sdgs, useCases, workflows, categoryTypes, search, sortColumn, sortDirection
  }
  const buildingBlockFilterDispatchValues = {
    setShowMature,
    setSDGs,
    setSdgs,
    setUseCases,
    setWorkflows,
    setCategoryTypes,
    setSearch,
    setSortColumn,
    setSortDirection
  }

  return (
    <BuildingBlockFilterContext.Provider value={buildingBlockFilterValues}>
      <BuildingBlockFilterDispatchContext.Provider value={buildingBlockFilterDispatchValues}>
        {children}
      </BuildingBlockFilterDispatchContext.Provider>
    </BuildingBlockFilterContext.Provider>
  )
}

export { BuildingBlockFilterProvider, BuildingBlockFilterContext, BuildingBlockFilterDispatchContext }
