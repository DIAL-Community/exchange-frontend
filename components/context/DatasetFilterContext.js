import { createContext, useState } from 'react'

const DatasetFilterContext = createContext()
const DatasetFilterDispatchContext = createContext()

const DatasetFilterProvider = ({ children }) => {
  const [origins, setOrigins] = useState([])
  const [countries, setCountries] = useState([])
  const [sectors, setSectors] = useState([])
  const [organizations, setOrganizations] = useState([])
  const [sdgs, setSdgs] = useState([])
  const [tags, setTags] = useState([])
  const [datasetTypes, setDatasetTypes] = useState([])

  const [search, setSearch] = useState('')

  const [pageNumber, setPageNumber] = useState(0)
  const [pageOffset, setPageOffset] = useState(0)

  const datasetFilterValues = {
    origins,
    countries,
    sectors,
    organizations,
    sdgs,
    tags,
    datasetTypes,
    search,
    pageOffset,
    pageNumber
  }
  const datasetFilterDispatchValues = {
    setOrigins,
    setCountries,
    setSectors,
    setOrganizations,
    setSdgs,
    setTags,
    setDatasetTypes,
    setSearch,
    setPageNumber,
    setPageOffset
  }

  return (
    <DatasetFilterContext.Provider value={datasetFilterValues}>
      <DatasetFilterDispatchContext.Provider value={datasetFilterDispatchValues}>
        {children}
      </DatasetFilterDispatchContext.Provider>
    </DatasetFilterContext.Provider>
  )
}

export { DatasetFilterProvider, DatasetFilterContext, DatasetFilterDispatchContext }
