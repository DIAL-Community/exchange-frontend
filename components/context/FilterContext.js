import { createContext, useEffect, useState } from 'react'
import { useRouter } from 'next/router'

const FilterContext = createContext()
const FilterDispatchContext = createContext()

const FilterProvider = ({ children }) => {
  const [search, setSearch] = useState('')
  // Task tracker context only
  const [showFailedOnly, setShowFailedOnly] = useState(false)
  const [showGovStackOnly, setShowGovStackOnly] = useState(false)
  const [showMature, setShowMature] = useState(false)

  const [categoryTypes, setCategoryTypes] = useState([])
  const [sdgs, setSdgs] = useState([])
  const [useCases, setUseCases] = useState([])
  const [workflows, setWorkflows] = useState([])

  const [origins, setOrigins] = useState([])
  const [countries, setCountries] = useState([])
  const [sectors, setSectors] = useState([])
  const [organizations, setOrganizations] = useState([])
  const [tags, setTags] = useState([])
  const [datasetTypes, setDatasetTypes] = useState([])

  const [aggregators, setAggregators] = useState([])
  const [operators, setOperators] = useState([])
  const [services, setServices] = useState([])

  const [years, setYears] = useState([])

  const [products, setProducts] = useState([])

  const [buildingBlocks, setBuildingBlocks] = useState([])
  const [showClosed, setShowClosed] = useState(false)

  const [aggregator, setAggregator] = useState(false)
  const [endorser, setEndorser] = useState(false)
  const [endorserLevel, setEndorserLevel] = useState('')

  const [specialties, setSpecialties] = useState([])
  const [certifications, setCertifications] = useState([])

  const [isEndorsed, setIsEndorsed] = useState(false)
  const [endorsers, setEndorsers] = useState([])
  const [licenseTypes, setLicenseTypes] = useState([])
  const [isLinkedWithDpi, setIsLinkedWithDpi] = useState(false)
  const [showDpgaOnly, setShowDpgaOnly] = useState(false)

  const [softwareCategories, setSoftwareCategories] = useState([])
  const [softwareFeatures, setSoftwareFeatures] = useState([])
  const [comparedProducts, setComparedProducts] = useState([])
  const [productStage, setProductStage] = useState(null)

  const [showBeta, setShowBeta] = useState(false)

  const router = useRouter()
  useEffect(() => {
    const handleRouteChange = (url) => {
      // Transitioning to other pages (pathname changing).
      if (url.indexOf(router.pathname) < 0) {
        setSearch('')
        setShowFailedOnly(false)
        setShowGovStackOnly(false)
        setShowMature(false)

        setCategoryTypes([])
        setSdgs([])
        setUseCases([])
        setWorkflows([])

        setOrigins([])
        setCountries([])
        setSectors([])
        setOrganizations([])
        setTags([])
        setDatasetTypes([])

        setAggregators([])
        setOperators([])
        setServices([])

        setYears([])

        setProducts([])

        setBuildingBlocks([])
        setShowClosed(false)

        setAggregator(false)
        setEndorser(false)
        setEndorserLevel('')

        setSpecialties([])
        setCertifications([])

        setIsEndorsed(false)
        setEndorsers([])
        setLicenseTypes([])
        setIsLinkedWithDpi(false)
        setShowDpgaOnly(false)

        setSoftwareCategories([])
        setSoftwareFeatures([])
        setComparedProducts([])
        setProductStage('')

        setShowBeta(false)
      }
    }

    router.events.on('routeChangeComplete', handleRouteChange)

    return () => {
      router.events.off('routeChangeComplete', handleRouteChange)
    }
  }, [router.events, router.asPath, router.pathname])

  const valueProps = {
    search,

    showFailedOnly,
    showGovStackOnly,
    showMature,
    categoryTypes,
    sdgs,
    useCases,
    workflows,

    countries,
    datasetTypes,
    organizations,
    origins,
    sectors,
    tags,

    aggregators,
    operators,
    services,

    years,

    products,

    buildingBlocks,
    showClosed,

    aggregator,
    endorser,
    endorserLevel,
    specialties,
    certifications,

    isEndorsed,
    endorsers,
    licenseTypes,
    isLinkedWithDpi,
    showDpgaOnly,
    softwareCategories,
    softwareFeatures,
    comparedProducts,
    productStage,

    showBeta
  }

  const dispatchProps = {
    setSearch,
    setShowFailedOnly,
    setShowGovStackOnly,
    setShowMature,
    setCategoryTypes,
    setSdgs,
    setUseCases,
    setWorkflows,

    setCountries,
    setDatasetTypes,
    setOrganizations,
    setOrigins,
    setSectors,
    setTags,

    setAggregators,
    setOperators,
    setServices,

    setYears,

    setProducts,

    setBuildingBlocks,
    setShowClosed,

    setAggregator,
    setEndorser,
    setEndorserLevel,
    setSpecialties,
    setCertifications,

    setIsEndorsed,
    setEndorsers,
    setLicenseTypes,
    setIsLinkedWithDpi,
    setShowDpgaOnly,
    setSoftwareCategories,
    setSoftwareFeatures,
    setComparedProducts,
    setProductStage,

    setShowBeta
  }

  return (
    <FilterContext.Provider value={{ ...valueProps }}>
      <FilterDispatchContext.Provider value={{ ...dispatchProps }}>
        {children}
      </FilterDispatchContext.Provider>
    </FilterContext.Provider>
  )
}

export { FilterProvider, FilterContext, FilterDispatchContext }
