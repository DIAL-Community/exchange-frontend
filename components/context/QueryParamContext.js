import { createContext, useState } from 'react'

const QueryParamContext = createContext()
const QueryParamContextProvider = ({ children }) => {
  const [interactionDetected, setInteractionDetected] = useState(false)
  return (
    <QueryParamContext.Provider value={{ ...{ interactionDetected, setInteractionDetected } }}>
      {children}
    </QueryParamContext.Provider>
  )
}

export { QueryParamContextProvider, QueryParamContext }
