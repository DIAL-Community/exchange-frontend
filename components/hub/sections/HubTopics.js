import { FormattedMessage } from 'react-intl'
import HubTopicTile from '../fragments/HubTopicTile'

const HubTopics = () => {

  return (
    <div className='flex flex-col gap-6 pb-12 max-w-catalog'>
      <img className='h-32 w-full object-cover' alt='DIAL Resource Hub' src='/images/hero-image/hub-hero.svg' />
      <div className='absolute w-full left-1/2 -translate-x-1/2' style={{ top: 'var(--ui-header-height)' }}>
        <div className='max-w-catalog mx-auto py-12'>
          <div className='text-2xl px-4 lg:px-8 xl:px-56 text-dial-gray'>
            Browse Resources by Topic
          </div>
        </div>
      </div>
      <div className='px-4 lg:px-8 xl:px-56 min-h-[60vh] lg:min-h-[65vh]'>
        <div className='flex flex-col gap-6 z-20'>
          <div className='max-w-4xl'>
            <FormattedMessage id='hub.topic.subtitle' />
          </div>
          <HubTopicTile />
        </div>
      </div>
    </div>
  )
}

export default HubTopics
