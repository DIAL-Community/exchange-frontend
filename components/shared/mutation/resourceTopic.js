import { gql } from '@apollo/client'

export const DELETE_RESOURCE_TOPIC = gql`
  mutation DeleteResourceTopic($id: ID!) {
    deleteResourceTopic(id: $id) {
      resourceTopic {
        id
      }
      errors
    }
  }
`

export const CREATE_RESOURCE_TOPIC = gql`
  mutation CreateResourceTopic(
    $name: String!
    $slug: String!
    $imageFile: Upload
    $description: String
    $parentTopicId: ID
  ) {
    createResourceTopic(
      name: $name
      slug: $slug
      imageFile: $imageFile
      description: $description
      parentTopicId: $parentTopicId
    ) {
      resourceTopic {
        id
        name
        slug
        imageFile
        parentTopic {
          id
          name
          slug
        }
        resourceTopicDescription {
          id
          description
          locale
        }
      }
      errors
    }
  }
`
