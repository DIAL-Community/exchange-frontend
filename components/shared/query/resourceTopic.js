import { gql } from '@apollo/client'

export const RESOURCE_TOPIC_SEARCH_QUERY = gql`
  query ResourceTopics($search: String) {
    resourceTopics(search: $search) {
      id
      name
      slug
    }
  }
`

export const RESOURCE_TOPIC_PAGINATION_ATTRIBUTES_QUERY = gql`
  query PaginationAttributeResourceTopic($search: String) {
    paginationAttributeResourceTopic(search: $search) {
      totalCount
    }
  }
`

export const PAGINATED_RESOURCE_TOPICS_QUERY = gql`
  query PaginatedResourceTopics(
    $search: String
    $limit: Int!
    $offset: Int!
  ) {
    paginatedResourceTopics(
      search: $search
      offsetAttributes: { limit: $limit, offset: $offset }
    ) {
      id
      name
      slug
      resourceTopicDescription {
        id
        description
        locale
      }
      resources {
        id
      }
    }
  }
`

export const RESOURCE_TOPIC_DETAIL_QUERY = gql`
  query ResourceTopic($slug: String!) {
    resourceTopic(slug: $slug) {
      id
      name
      slug
      resourceTopicDescription {
        id
        description
        locale
      }
      resources {
        id
        name
        slug
        imageFile
      }
    }
  }
`
