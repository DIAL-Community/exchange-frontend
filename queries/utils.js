const MIN_SEARCH_VALUE_CHARACTERS = 2

export const fetchSelectOptions = async (client, input, query, fetchedDataCallback, locale = null) => {
  const variables = {
    search: input
  }

  if (input && input.trim().length < MIN_SEARCH_VALUE_CHARACTERS) {
    return []
  } else if (locale) {
    variables.locale = locale
  }

  const response = await client.query({ query, variables })

  return response.data ? fetchedDataCallback(response.data) : []
}

export const fetchSelectOptionsWithMature = async (client, input, query, fetchedDataCallback, locale = null) => {
  const variables = {
    search: input,
    mature: true
  }

  if (input && input.trim().length < MIN_SEARCH_VALUE_CHARACTERS) {
    return []
  } else if (locale) {
    variables.locale = locale
  }

  const response = await client.query({ query, variables })

  return response.data ? fetchedDataCallback(response.data) : []
}
