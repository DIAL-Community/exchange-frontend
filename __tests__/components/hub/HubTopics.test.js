import { screen } from '@testing-library/dom'
import { FilterContextProvider } from '../../../components/context/FilterContext'
import { QueryParamContextProvider } from '../../../components/context/QueryParamContext'
import HubTopics from '../../../components/hub/sections/HubTopics'
import { RESOURCE_TOPIC_SEARCH_QUERY } from '../../../components/shared/query/resourceTopic'
import { render } from '../../test-utils'
import CustomMockedProvider, { generateMockApolloData } from '../../utils/CustomMockedProvider'
import { mockNextUseRouter, mockTenantApi } from '../../utils/nextMockImplementation'
import { resourceTopics } from './data/HubTopics.data'

mockTenantApi()
mockNextUseRouter()
describe('Unit tests for the opportunity detail page.', () => {
  const mockResourceTiles = generateMockApolloData(
    RESOURCE_TOPIC_SEARCH_QUERY,
    {
      'search': ''
    },
    null,
    resourceTopics
  )

  test('Should render detail of a topic page.', async () => {
    const { container } = render(
      <CustomMockedProvider mocks={[mockResourceTiles]}>
        <QueryParamContextProvider>
          <FilterContextProvider>
            <HubTopics />
          </FilterContextProvider>
        </QueryParamContextProvider>
      </CustomMockedProvider>
    )

    expect(await screen.findByText('AI + Machine Learning')).toBeInTheDocument()
    expect(await screen.findByText('Consent')).toBeInTheDocument()
    expect(await screen.findByText('Data Privacy')).toBeInTheDocument()
    expect(await screen.findByText('Data Exchange')).toBeInTheDocument()
    expect(container).toMatchSnapshot()
  })
})
