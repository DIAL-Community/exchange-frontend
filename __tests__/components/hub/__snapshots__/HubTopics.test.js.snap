// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Unit tests for the opportunity detail page. Should render detail of a topic page. 1`] = `
<div>
  <div
    class="Toastify"
  />
  <div
    class="flex flex-col gap-6 pb-12 max-w-catalog"
  >
    <img
      alt="DIAL Resource Hub"
      class="h-32 w-full object-cover"
      src="/images/hero-image/hub-hero.svg"
    />
    <div
      class="absolute w-full left-1/2 -translate-x-1/2"
    >
      <div
        class="max-w-catalog mx-auto py-12"
      >
        <div
          class="text-2xl px-4 lg:px-8 xl:px-56 text-dial-gray"
        >
          Browse Resources by Topic
        </div>
      </div>
    </div>
    <div
      class="px-4 lg:px-8 xl:px-56 min-h-[60vh] lg:min-h-[65vh]"
    >
      <div
        class="flex flex-col gap-6 z-20"
      >
        <div
          class="max-w-4xl"
        >
          
    Topics showcases a series of knowledge products published by DIAL and its partners
    on key issues surrounding digital public infrastructure.
  
        </div>
        <div
          class="grid grid-cols-1 md:grid-cols-2 gap-12 mx-auto"
        >
          <a
            class="text-dial-cotton"
            href="/hub/topics/ai--machine-learning"
          >
            <div
              class="bg-dial-deep-purple py-8 lg:py-12 w-52 h-56 relative flex justify-center"
            >
              <img
                alt="AI + Machine Learning"
                class="white-filter opacity-50 mb-10 h-16"
                src="http://localhost:3000//assets/resources/resource-topic-placeholder.svg"
              />
              <div
                class="absolute left-1/2 -translate-x-1/2 bottom-8 text-center uppercase 2xl:text-lg"
              >
                AI + Machine Learning
              </div>
            </div>
          </a>
          <a
            class="text-dial-cotton"
            href="/hub/topics/consent"
          >
            <div
              class="bg-dial-deep-purple py-8 lg:py-12 w-52 h-56 relative flex justify-center"
            >
              <img
                alt="Consent"
                class="white-filter opacity-50 mb-10 h-16"
                src="http://localhost:3000//assets/resources/resource-topic-placeholder.svg"
              />
              <div
                class="absolute left-1/2 -translate-x-1/2 bottom-8 text-center uppercase 2xl:text-lg"
              >
                Consent
              </div>
            </div>
          </a>
          <a
            class="text-dial-cotton"
            href="/hub/topics/data-privacy"
          >
            <div
              class="bg-dial-deep-purple py-8 lg:py-12 w-52 h-56 relative flex justify-center"
            >
              <img
                alt="Data Privacy"
                class="white-filter opacity-50 mb-10 h-16"
                src="http://localhost:3000//assets/resources/resource-topic-placeholder.svg"
              />
              <div
                class="absolute left-1/2 -translate-x-1/2 bottom-8 text-center uppercase 2xl:text-lg"
              >
                Data Privacy
              </div>
            </div>
          </a>
          <a
            class="text-dial-cotton"
            href="/hub/topics/data-exchange"
          >
            <div
              class="bg-dial-deep-purple py-8 lg:py-12 w-52 h-56 relative flex justify-center"
            >
              <img
                alt="Data Exchange"
                class="white-filter opacity-50 mb-10 h-16"
                src="http://localhost:3000//assets/resources/resource-topic-placeholder.svg"
              />
              <div
                class="absolute left-1/2 -translate-x-1/2 bottom-8 text-center uppercase 2xl:text-lg"
              >
                Data Exchange
              </div>
            </div>
          </a>
        </div>
      </div>
    </div>
  </div>
</div>
`;
